### DISCUSSION TITLE
AD Neumann BC for a vector variable

### DISCUSSION THREAD

**Discussion by ABallisat** [idaholab/moose#17018] [Q&A Modules: General] [answered]
Hello all,

I have a need to set a Neumann BC for a vector variable, specifically setting a component of the gradient of a component of a vector to be a value (probably zero), for example something along the lines of
```
partial v_x / partial y = 0
```
for vector v with components x and y (and naturally extend to 3D). I am using AD so would prefer it to be within that system. I have been looking at the ADIntegratedBC.C as a basis for this but cannot figure out how to get it to work with vector variables. Could someone point me in the right direction? 

I have looked at VectorNeumannBC.C but cannot figure out how it iterates over a vector variable, my understanding in this area is lacking. My guess is that the 
```C
for (_i = 0; _i < _test.size(); _i++)
```
line iterates over the components as _test.size() returns the number of components of the vector, is this understanding correct? If so, then i could be passed through to work out what component you are on. 

Thanks for the help!

Alex

**Comment by GiudGiud** [comment, marked as answer]
Hi Alex

`VectorNeumannBC` are definitely a good start to create an AD version of that BC.
`_i` will indeed be the component the BC is currently considering.

You may want to look at `ADVectorDirichletBC` as well btw.

Guillaume
